# **SUPERVISED LEARNING**

**AI05123 - VIRGIE YUNITA SALSABIL**

# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
# Tampilan dataset
df

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0             1    15634602    France  Female   42       2       0.00   
1             2    15647311     Spain  Female   41       1   83807.86   
2             3    15619304    France  Female   42       8  159660.80   
3             4    15701354    France  Female   39       1       0.00   
4             5    15737888     Spain  Female   43       2  125510.82   
...         ...         ...       ...     ...  ...     ...        ...   
9995       9996    15606229    France    Male   39       5       0.00   
9996       9997    15569892    France    Male   35      10   57369.61   
9997       9998    15584532    France  Female   36       7       0.00   
9998       9999    15682355   Germany    Male   42       3   75075.31   
9999      10000    15628319    France  Female   28       4  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0                 1          1               1        101348.88       1  
1                 1          0               1        112542.58       0  
2                 3          1               0        113931.57       1  
3                 2          0               0         93826.63       0  
4                 1          1               1         79084.10       0  
...             ...        ...             ...              ...     ...  
9995              2          1               0         96270.64       0  
9996              1          1               1        101699.77       0  
9997              1          0               1         42085.58       1  
9998              2          1               0         92888.52       1  
9999              1          1               0         38190.78       0  

[10000 rows x 12 columns]

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited']].copy()

df

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0       France  Female   42       2       0.00              1          1   
1        Spain  Female   41       1   83807.86              1          0   
2       France  Female   42       8  159660.80              3          1   
3       France  Female   39       1       0.00              2          0   
4        Spain  Female   43       2  125510.82              1          1   
...        ...     ...  ...     ...        ...            ...        ...   
9995    France    Male   39       5       0.00              2          1   
9996    France    Male   35      10   57369.61              1          1   
9997    France  Female   36       7       0.00              1          0   
9998   Germany    Male   42       3   75075.31              2          1   
9999    France  Female   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  
0                  1        101348.88       1  
1                  1        112542.58       0  
2                  0        113931.57       1  
3                  0         93826.63       0  
4                  1         79084.10       0  
...              ...              ...     ...  
9995               0         96270.64       0  
9996               1        101699.77       0  
9997               1         42085.58       1  
9998               0         92888.52       1  
9999               0         38190.78       0  

[10000 rows x 10 columns]

In [ ]:
#Menampilkan kolom pada df yg bertipe objek/kategorik

categorical = df.select_dtypes(include=['object']).columns
categorical

Index(['Geography', 'Gender'], dtype='object')

## Preprocessing

In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'])

In [ ]:
# Tampilan df setelah dilakukan One-Hot Encoder data categorical
df

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0      42       2       0.00              1          1               1   
1      41       1   83807.86              1          0               1   
2      42       8  159660.80              3          1               0   
3      39       1       0.00              2          0               0   
4      43       2  125510.82              1          1               1   
...   ...     ...        ...            ...        ...             ...   
9995   39       5       0.00              2          1               0   
9996   35      10   57369.61              1          1               1   
9997   36       7       0.00              1          0               1   
9998   42       3   75075.31              2          1               0   
9999   28       4  130142.79              1          1               0   

      EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0           101348.88       1                 1                  0   
1           112542.58       0                 0                  0   
2           113931.57       1                 1                  0   
3            93826.63       0                 1                  0   
4            79084.10       0                 0                  0   
...               ...     ...               ...                ...   
9995         96270.64       0                 1                  0   
9996        101699.77       0                 1                  0   
9997         42085.58       1                 1                  0   
9998         92888.52       1                 0                  1   
9999         38190.78       0                 1                  0   

      Geography_Spain  Gender_Female  Gender_Male  
0                   0              1            0  
1                   1              1            0  
2                   0              1            0  
3                   0              1            0  
4                   1              1            0  
...               ...            ...          ...  
9995                0              0            1  
9996                0              0            1  
9997                0              1            0  
9998                0              0            1  
9999                0              1            0  

[10000 rows x 13 columns]

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop(columns=['Exited'])  # Mengambil semua kolom kecuali 'Exited' sebagai feature
y = df['Exited']  # Mengambil kolom 'Exited' sebagai target

In [ ]:
#Lakukan Scaler dan/atau Normalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE

#Inisialisasi MinMaxScaler
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisasi. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)

#X_transform = X.copy()

In [ ]:
# Tampilan X setelah dilakukan scaler
X_transform

Age  Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     0.324324     0.2  0.000000       0.000000        1.0             1.0   
1     0.310811     0.1  0.334031       0.000000        0.0             1.0   
2     0.324324     0.8  0.636357       0.666667        1.0             0.0   
3     0.283784     0.1  0.000000       0.333333        0.0             0.0   
4     0.337838     0.2  0.500246       0.000000        1.0             1.0   
...        ...     ...       ...            ...        ...             ...   
9995  0.283784     0.5  0.000000       0.333333        1.0             0.0   
9996  0.229730     1.0  0.228657       0.000000        1.0             1.0   
9997  0.243243     0.7  0.000000       0.000000        0.0             1.0   
9998  0.324324     0.3  0.299226       0.333333        1.0             0.0   
9999  0.135135     0.4  0.518708       0.000000        1.0             0.0   

      EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0            0.506735               1.0                0.0              0.0   
1            0.562709               0.0                0.0              1.0   
2            0.569654               1.0                0.0              0.0   
3            0.469120               1.0                0.0              0.0   
4            0.395400               0.0                0.0              1.0   
...               ...               ...                ...              ...   
9995         0.481341               1.0                0.0              0.0   
9996         0.508490               1.0                0.0              0.0   
9997         0.210390               1.0                0.0              0.0   
9998         0.464429               0.0                1.0              0.0   
9999         0.190914               1.0                0.0              0.0   

      Gender_Female  Gender_Male  
0               1.0          0.0  
1               1.0          0.0  
2               1.0          0.0  
3               1.0          0.0  
4               1.0          0.0  
...             ...          ...  
9995            0.0          1.0  
9996            0.0          1.0  
9997            1.0          0.0  
9998            0.0          1.0  
9999            1.0          0.0  

[10000 rows x 12 columns]

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model 1 : Logistic Regression

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model
    #from sklearn.linear_model import LogisticRegression
    #model1 = LogisticRegression()
    #params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

# Import library yang diperlukan dari scikit-learn
from sklearn.linear_model import LogisticRegression

#START CODE

# Inisialisasi model logistik regresi
model1 = LogisticRegression()
# Buat kumpulan Hyperparameter yg akan diuji
params =  {
    'tol': [0.0001, 0.001, 0.01, 0.1],  # Toleransi (tolerance)
    'C': [0.001, 0.01, 0.1, 1, 10, 100]  # Kebalikan kekuatan regulasi
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1, # Model yang akan diuji
             param_grid= params, # Kumpulan hyperparameter yang akan diuji
             scoring = 'accuracy', # Metrik evaluasi yang digunakan
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 10-fold cross validation (artinya kita melakukan iterasi model sebanyak 10 kali)
            )

# Melatih model dengan kombinasi hyperparameter terbaik
grid.fit(X_train,y_train)

# Menampilkan hyperparameter terbaik yang ditemukan oleh GridSearchCV
best_params = grid.best_params_
print("Hyperparameter terbaik:", best_params)

Hyperparameter terbaik: {'C': 0.1, 'tol': 0.0001}


In [ ]:
# Lakukan evaluasi dengan beberapa metric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

# Prediksi label dari data uji menggunakan model yang telah dituning
y_pred = grid.predict(X_test)

# Menampilkan laporan klasifikasi
print("Classification Report:")
print(classification_report(y_test, y_pred))
print("")

# Menampilkan Confusion Matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("")

# Menampilkan akurasi model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.98      0.89      1983
           1       0.72      0.15      0.25       517

    accuracy                           0.81      2500
   macro avg       0.77      0.57      0.57      2500
weighted avg       0.80      0.81      0.76      2500


Confusion Matrix:
[[1953   30]
 [ 439   78]]

Accuracy: 0.8124


**PENJELASAN**

- Logistic Regression digunakan ketika kita ingin melakukan klasifikasi, seperti memprediksi apakah suatu observasi masuk ke dlm salah satu dari dua kelas (binary classification).

- Disini kita melakukan parameter tuning dg  GridSearchCV utk mencari kombinasi terbaik dari hyperparameter 'C' dan 'tol' yg meminimalkan error pada model Logistic Regression.

- Hasil akurasi model sekitar 0.8124, yang berarti model ini memiliki tingkat akurasi yang cukup baik dalam memprediksi label kelas.

- Model ini memiliki akurasi yg cukup baik dlm memprediksi kelas 0, tetapi memiliki masalah dlm memprediksi kelas 1, yang ditunjukkan oleh recall yang rendah dan F1-score yang rendah untuk kelas 1.

## Model 2 : Random Forest Classifier
Jelaskan secara Singkat Model ke-2 yang digunakan!

In [ ]:
from sklearn.ensemble import RandomForestClassifier

#START CODE

# Inisialisasi model Random Forest
model2 = RandomForestClassifier()

# Definisikan kumpulan hyperparameter yang akan diuji untuk model Random Forest
params = {
    'n_estimators': [50, 100, 200],  # Jumlah pohon dalam ensemble
    'max_depth': [5, 10, 20, 30],  # Kedalaman maksimum setiap pohon
    'min_samples_split': [2, 5, 10],  # Jumlah sampel minimum untuk membagi simpul internal
    'min_samples_leaf': [1, 2, 4],  # Jumlah sampel minimum di setiap daun pohon
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model2, # Model yang akan diuji
             param_grid= params, # Kumpulan hyperparameter yang akan diuji
             scoring = 'accuracy', # Metrik evaluasi yang digunakan
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 10-fold cross validation (artinya kita melakukan iterasi model sebanyak 10 kali)
            )

# Melatih model Random Forest dengan kombinasi hyperparameter terbaik
grid.fit(X_train, y_train)

# Menampilkan hyperparameter terbaik yang ditemukan oleh GridSearchCV untuk model Random Forest
best_params = grid.best_params_
print("Hyperparameter terbaik untuk Random Forest:", best_params)

Hyperparameter terbaik untuk Random Forest: {'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 50}


In [ ]:
# Lakukan evaluasi dengan beberapa metric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

# Prediksi label dari data uji menggunakan model yang telah dituning
y_pred = grid.predict(X_test)

# Menampilkan laporan klasifikasi
print("Classification Report:")
print(classification_report(y_test, y_pred))
print("")

# Menampilkan Confusion Matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("")

# Menampilkan akurasi model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.80      0.44      0.57       517

    accuracy                           0.86      2500
   macro avg       0.84      0.71      0.75      2500
weighted avg       0.86      0.86      0.85      2500


Confusion Matrix:
[[1926   57]
 [ 287  230]]

Accuracy: 0.8624


**PENJELASAN**

- Model Random Forest yaitu Model ensemble yang menggabungkan beberapa pohon keputusan (decision trees) utk meningkatkan kinerja prediksi.

- Hyperparameter yg dgunakan = jmlh pohon (n_estimators), kedalaman maks setiap pohon (max_depth), jmlh sampel min utk membagi simpul internal (min_samples_split), dan jmlh sampel minimum di setiap daun pohon (min_samples_leaf).

- Parameter terbaik yang ditentukan memiliki akurasi yang lebih baik yaitu  0.8624 daripada model sebelumnya (Logistic Regression).

- Namun, model ini masih memiliki recall yang rendah untuk kelas 1 (terpenuhi), yang berarti masih perlu perbaikan dalam mengidentifikasi kasus positif kelas 1.


## Model 3 : Support Vector Machine (SVM)
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

In [ ]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

#START CODE
# Inisialisasi model Support Vector Machine
model3 = SVC()

# Definisikan kumpulan hyperparameter yang akan diuji
params = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],  # Parameter kebalikan regulasi
    'kernel': ['rbf', 'poly', 'sigmoid'],  # Jenis kernel
    'gamma': [0.001, 0.01, 0.1, 1, 'scale', 'auto'],  # Parameter kernel
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model3, # Model yang akan diuji
             param_grid= params, # Kumpulan hyperparameter yang akan diuji
             scoring = 'accuracy', # Metrik evaluasi yang digunakan
             n_jobs = 10, # core cpu yang digunakan
             cv = 3 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )
# Melatih model dengan kombinasi hyperparameter terbaik
grid.fit(X_train, y_train)

# Menampilkan hyperparameter terbaik yang ditemukan oleh GridSearchCV
best_params = grid.best_params_
print("Hyperparameter terbaik:", best_params)

Hyperparameter terbaik: {'C': 100, 'gamma': 'scale', 'kernel': 'poly'}


In [ ]:
# Lakukan evaluasi dengan beberapa metric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

# Prediksi label dari data uji menggunakan model yang telah dituning
y_pred = grid.predict(X_test)

# Menampilkan laporan klasifikasi
print("Classification Report:")
print(classification_report(y_test, y_pred))
print("")

# Menampilkan Confusion Matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("")

# Menampilkan akurasi model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.98      0.92      1983
           1       0.82      0.44      0.57       517

    accuracy                           0.86      2500
   macro avg       0.85      0.71      0.75      2500
weighted avg       0.86      0.86      0.85      2500


Confusion Matrix:
[[1934   49]
 [ 289  228]]

Accuracy: 0.8648


**PENJELASAN**

-  SVM digunakan utk mencari hyperplane terbaik yg memisahkan dua kelas dalam ruang fitur.

- Hyperparameter yg dgunakan =
<br> **C :** Nilai yang lebih tinggi mengarah pada margin yang lebih ketat.
<br> **kernel :** beberapa jenis kernel yg diuji -> 'rbf' (Radial Basis Function), 'poly' (Polynomial), dan 'sigmoid'.
<br> **gamma :** Parameter kernel yang mengatur seberapa jauh pengaruh dari satu titik data terhadap yang lain dalam fungsi kernel.

- Model SVM memiliki akurasi yg baik (0.8648) dlm memprediksi kelas mayoritas (kelas 0), tetapi memiliki performa yang lebih rendah dalam memprediksi kelas minoritas (kelas 1), seperti yang tercermin dari nilai recall dan F1-score yang lebih rendah untuk kelas 1.


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

- ***Logistic Regression:***

Akurasi: 0.8124 <br>
Recall untuk kelas 1: 0.15 <br>
F1-score untuk kelas 1: 0.25

- ***Random Forest:***

Akurasi: 0.8624 <br>
Recall untuk kelas 1: 0.44 <br>
F1-score untuk kelas 1: 0.57

- ***Support Vector Machine (SVM):***

Akurasi: 0.8648 <br>
Recall untuk kelas 1: 0.44 <br>
F1-score untuk kelas 1: 0.57


- Dilihat dari akurasi, model Random Forest dan SVM memiliki kinerja yang lebih baik daripada Logistic Regression.

- Tetapi, semua model di atas memiliki recall yang rendah untuk kelas 1 (terpenuhi), yang berarti mereka kurang baik dalam mengidentifikasi kasus positif kelas 1. Meskipun Recall untuk kelas 1 sama di model Random Forest dan SVM, F1-score untuk kelas 1 juga sama, dan menunjukkan kinerja yang serupa.

- Berdasarkan akurasi dan kinerja pada kelas 1, tidak ada perbedaan besar antara model Random Forest dan SVM.

**KESIMPULAN**

Jadi, model yg dapat dikatakan paling baik adalah model ***Support Vector Machine (SVM)*** berdasarkan hasil akurasi yg lebih tinggi dibanding model lainnya yaitu **0.8648**. Namun, perlu diingat bahwa pemilihan model terbaik harus mempertimbangkan berbagai faktor seperti kompleksitas model, waktu pelatihan, dan interpretasi hasil.